### DISCUSSION TITLE
bcom vs coverap: discrepancy

### DISCUSSION THREAD

**Discussion by kgaalswyk** [GrossfieldLab/loos#83] [Q&A] [answered]
Hi all,

We've been using the bcom tool for a while on a few systems and wanted to do a bit of a sanity check by making some comparisons to what the coverlap tool provides. Based on my understanding of these two papers ([DOI: 10.1103/PhysRevE.65.031910](https://journals.aps.org/pre/pdf/10.1103/PhysRevE.65.031910), [DOI:10.1021/ct2002754](https://pubs.acs.org/doi/full/10.1021/ct2002754)) if we compute the covariance overlap (using coverlap) of the 1st half of the trajectory with the full trajectory and average this with the covariance overlap of the 2nd half of the trajectory with the full trajectory, then this value should (in principle) be the same as if (using bcom) we have a block size of NFRAMES/2 (half the trajectory length) as this results in 2 blocks whose overlap is being compared with the full trajectory and then averaged together.

Assuming this, we ran bcom using the following command:
`bcom -s '!hydrogen' --blocks 50:50:30000 [prot].prmtop [prot].dcd > [prot]_bcom.out`

We then extract the line where there are 2 blocks each 1/2 of the trajectory (in this case 14950 frames as we're skipping the first 50). For a sample protein this value is 0.896.

Then to test coverlap we first generate the PCA matrices using svd for 1st 1/2, 2nd 1/2, and full:
`svd -A '!hydrogen' -S '!hydrogen' -r [0:14999] -p [prot]_1 [prot].prmtop [prot].dcd`
`svd -A '!hydrogen' -S '!hydrogen' -r [15000:29999] -p [prot]_2 [prot].prmtop [prot].dcd`
`svd -A '!hydrogen' -S '!hydrogen' -r [0:29999] -p [prot]_full [prot].prmtop [prot].dcd`

Then run coverlap comparing both halves to the whole:
`coverlap [prot]_1_s.asc [prot]_1_U.asc [prot]_full_s.asc [prot]_full_U.asc > [prot]_1_coverlap.out`
`coverlap [prot]_2_s.asc [prot]_2_U.asc [prot]_full_s.asc [prot]_full_U.asc > [prot]_2_coverlap.out`

The average of these two covariance overlap values for a sample protein is 0.828.

My question is why is there this discrepancy? Is there something in the commands I'm missing? I looked through the code a bit and couldn't discern any flags/options that are being set for one that aren't being set for the other, and they both make use of the same "covarianceoverlap" command. So I assume this means that the input eigenvectors and eigenvalues are different, but I cannot figure out how or why. Or is there something inherently different in the processes that I didn't see?

Cheers,

Kari Gaalswyk
Postdoctoral Fellow
University of Denver

**Comment by agrossfield** [comment, marked as answer]
The only things that occur to me off the top of my head are round off error (unlikely to be that large, but it’s a lot of frames and if I recall correctly we do the calculation with floats) or something to do with the alignment.  

Of the two, I think the latter is more likely. I think the blocked pca uses the same alignment from the full trajectory alignment, rather than redoing it. In that case, the two half alignments are slightly different. You could test this by pre-aligning the trajectory with the `aligner` tool, then telling `svd` not to align. If my hypothesis is correct, that should make the discrepancy go away. In that case, it’s not a bug. 

@tromo , do you have any thoughts?

**Reply by tromo** [reply]
I think you nailed it with number 2 above...  `bcom` does a global alignment, but the blocks are not re-aligned.  You will need to run `aligned` to get the global alignment, then tell `svd` to _not_ align.

**Reply by kgaalswyk** [reply]
Ok. So I used the aligner tool to align the full trajectory and then ran svd and coverlap as shown below:
`aligner --prefix [prot]_aligned --align '!hydrogen' [prot].prmtop [prot].dcd`

`svd -S '!hydrogen' --noalign 1 -r 0:14999 -p [prot]_1_aligned [prot]_aligned.pdb [prot]_aligned.dcd`
`svd -S '!hydrogen' --noalign 1 -r 15000:29999 -p [prot]_2_aligned [prot]_aligned.pdb [prot]_aligned.dcd`
`svd -S '!hydrogen' --noalign 1 -r 0:29999 -p [prot]_aligned [prot]_aligned.pdb [prot]_aligned.dcd`

`coverlap [prot]_1_aligned_s.asc [prot]_1_aligned_U.asc [prot]_aligned_s.asc [prot]_aligned_U.asc > [prot]_1_coverlap.out`
`coverlap [prot]_2_aligned_s.asc [prot]_2_aligned_U.asc [prot]_aligned_s.asc [prot]_aligned_U.asc > [prot]_2_coverlap.out`

And while this reduces the difference, it doesn't eliminate it entirely. I ran this for 7 different trajectories (replicates of the same system) and the difference ranges from 0.0005 to 0.045 with neither always being greater than the other. For the system I mentioned earlier we now have a coverlap of 0.849 and a bcom of 0.896. I'd assume that if there were rounding issues that the difference would be a bit more regular, but perhaps not.

Best,
Kari Gaalswyk

**Reply by agrossfield** [reply]
My intuition is that what's left is numerical, but I'm not even sure how to test.

**Reply by tromo** [reply]
I agree.  In addition, there are under-the-hood differences that will contribute to discrepancies.  The svd tool uses dgesvd() to calculate the SVD, whereas bcom calls ssyev() to get an eigen decomp (and I believe the default is to normalize the eigenvalues by block size).  I would only expect to get "reasonably" close numbers, not identical ones (nor a systematic difference).

**Reply by agrossfield** [reply]
for what it's worth, those differences are also well below the threshold of interpretability.
